staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,584 @@
+			defer qs.lock.Unlock()
+
+			// TODO(aaron-prindle) add metrics to these two cases
+			if req.Enqueued {
+				// remove the request from the queue as it has timed out
+				for i := range req.Queue.Requests {
+					if req == req.Queue.Requests[i] {
+						// remove the request
+						req.Queue.Requests = append(req.Queue.Requests[:i],
+							req.Queue.Requests[i+1:]...)
+						break
+					}
+				}
+				// At this point, if the qs is quiescing,
+				// has zero requests executing, and has zero requests enqueued
+				// then a call to the EmptyHandler should be forked.
+				if qs.emptyHandler != nil && qs.numRequestsEnqueued == 0 &&
+					qs.getRequestsExecutingLocked() == 0 {
+					qs.counter.Add(1)
+					go func(eh fq.EmptyHandler) {
+						defer runtime.HandleCrash()
+						defer qs.counter.Add(-1)
+						eh.HandleEmpty()
+					}(qs.emptyHandler)
+				}
+			} else {
+				// At this point we know that req was in its queue earlier and another
+				// goroutine has removed req from its queue and called qs.counter.Add(1)
+				// in anticipation of unblocking this goroutine through the other arm of this
+				// select. In this case we need to decrement the counter because this goroutine
+				// was actually unblocked through a different code path.
+				qs.counter.Add(-1)
+			}
+		}()
+		return false, false, func() {}
+	}
+}
+
+// syncTimeLocked is used to sync the time of the queueSet by looking at the elapsed
+// time since the last sync and this value based on the 'virtualtime ratio'
+// which scales inversely to the # of active flows
+func (qs *queueSet) syncTimeLocked() {
+	realNow := qs.clock.Now()
+	timesincelast := realNow.Sub(qs.lastRealTime).Seconds()
+	qs.lastRealTime = realNow
+	var virtualTimeRatio float64
+
+	activeQueues := 0
+	reqs := 0
+	for _, queue := range qs.queues {
+		reqs += queue.RequestsExecuting
+
+		if len(queue.Requests) > 0 || queue.RequestsExecuting > 0 {
+			activeQueues++
+		}
+	}
+	// no active flows, virtual time does not advance (also avoid div by 0)
+	if activeQueues == 0 {
+		virtualTimeRatio = 0
+	} else {
+		virtualTimeRatio = math.Min(float64(reqs), float64(qs.config.ConcurrencyLimit)) / float64(activeQueues)
+	}
+
+	qs.virtualTime += timesincelast * virtualTimeRatio
+}
+
+func (qs *queueSet) lockAndSyncTime() {
+	qs.lock.Lock()
+	qs.syncTimeLocked()
+}
+
+// timeoutOldRequestsAndRejectOrEnqueue encapsulates the logic required
+// to validate and enqueue a request for the queueSet/QueueSet:
+// 1) Start with shuffle sharding, to pick a queue.
+// 2) Reject old requests that have been waiting too long
+// 3) Reject current request if there is not enough concurrency shares and
+// we are at max queue length
+// 4) If not rejected, create a request and enqueue
+// returns the enqueud request on a successful enqueue
+// returns nil in the case that there is no available concurrency or
+// the queuelengthlimit has been reached
+// NOTE: This function must only be called with the QueueSet locked
```

Comment:
Add the "Locked" suffix to the name, since you started doing that elsewhere?

Reply:
Updated - s/timeoutOldRequestsAndRejectOrEnqueue/timeoutOldRequestsAndRejectOrEnqueueLocked (added Locked suffix)